### PR TITLE
Add SDE Samplers: Curved Euler and Overshoot

### PR DESCRIPTION
**Pull request by Abhinay1997**
# What does this PR do?

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

Hello @yiyixuxu, this PR adds two SDE schedulers as described in [Improving Rectified Flow with Boundary Conditions](https://arxiv.org/abs/2506.15864) and [AMO Sampler](https://github.com/hxixixh/amo-release)

I have added the docs and the scheduler files. Let me know if I missed something.

Test code:

```python
import torch
from diffusers import FluxPipeline
from diffusers.schedulers import FlowMatchStochasticOvershootDiscreteScheduler, FlowMatchStochasticCurvedEulerScheduler

pipe = FluxPipeline.from_pretrained('black-forest-labs/FLUX.1-Krea-dev', torch_dtype=torch.bfloat16)
pipe.to('cuda')

euler_scheduler = pipe.scheduler
overshoot_scheduler = FlowMatchStochasticOvershootDiscreteScheduler.from_pretrained('black-forest-labs/FLUX.1-Krea-dev', subfolder='scheduler')
cec_scheduler = FlowMatchStochasticCurvedEulerScheduler.from_pretrained('black-forest-labs/FLUX.1-Krea-dev', subfolder='scheduler')

prompt = "A kangaroo holding a sign that says 'Sting like a bee, float like a butterfly' wearing black sunglasses and boxing gloves, the golden gate bridge in the background."

pipe.scheduler = euler_scheduler
out = pipe(
    prompt=prompt,
    height=1024,
    width=1024,
    guidance_scale=3.5,
    num_inference_steps=28,
    max_sequence_length=512,
    generator=torch.Generator("cpu").manual_seed(0)
).images[0]
out.save('euler.jpg')

## CIRCULAR EULER
pipe.scheduler = cec_scheduler
out = pipe(
    prompt=prompt,
    height=1024,
    width=1024,
    guidance_scale=3.5,
    num_inference_steps=28,
    max_sequence_length=512,
    generator=torch.Generator("cpu").manual_seed(0)
).images[0]
out.save('circular_euler.jpg')

## OVERSHOOT C = 1.0 default
pipe.scheduler = overshoot_scheduler
pipe.scheduler.set_c(2.0)
out = pipe(
    prompt=prompt,
    height=1024,
    width=1024,
    guidance_scale=3.5,
    num_inference_steps=28,
    max_sequence_length=512,
    generator=torch.Generator("cpu").manual_seed(0)
).images[0]
out.save('overshoot.jpg')
```

Attached comparisons with `FlowMatchEulerDiscreteScheduler` with `stochastic_sampling` set to `False` & `True`.

grid

grid

### CODE REVIEW

**Comment by Abhinay1997** [comment]
Would like to hear your thoughts as well, @a-r-r-o-w

**Comment by a-r-r-o-w** [comment]
Thanks for the amazing work @Abhinay1997! I can help do a full code review soon, but it looks good from a quick glance. Good writeup as well!

Always great to have more flexibility with new techniques if there's a principled study on it. The final decision for supporting new schedulers is @yiyixuxu's so let's wait for her review :) I'll try to understand the papers and check for correctness in the mean time

The generated images look good overall. Not a blocker on my end, but I can notice some block-ness and artifacts in the images. It's probably due to downscaling of the images when creating the grid, but I'll ping @asomoza for further exploration as his eyes are much better than mine in analyzing image quality.

**Comment by Abhinay1997** [comment]
Thank you for the kind words! Of course, I'll wait for her review and thoughts on this. As for the blockiness and artifacts, that does seem to be an export artifact as I rendered the grid in tldraw and exported the whole thing as as single low res image. 

But the smoothing does come from the SDE process itself which is worth the tradeoff of better text rendering and randomness maybe.

**Comment by Abhinay1997** [comment]
Gently pinging @yiyixuxu . Would love to hear your thoughts on this PR.
